### PR TITLE
Centralized SFSDKLogger code

### PR DESCRIPTION
**Pull request by khawkins**
For the team's consideration.  We have a lot of duplicated logger code in the new logging system, that seems ripe for centralizing through an inheritance model.  This code implements that approach with `SFSDKAnalyticsLogger`.  If this is agreeable, I can pretty quickly bang out the rest of the logging to follow suit.

### CODE REVIEW

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h [hunk 1]
```diff
@@ -94,6 +102,15 @@
 - (void)e:(nonnull Class)class message:(nonnull NSString *)message;
 
 /**
+ * Logs an error log line.
+ *
+ * @param class Class.
+ * @param format The format message, and optional arguments to expand in the format.
+ * @param ... The arguments to the message format string.
+ */
+- (void)e:(nonnull Class)class format:(nonnull NSString *)format, ...;
```
Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: These are just added for symmetry.  Seems like both method signatures should exist for each log level.

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h [hunk 2]
```diff
@@ -144,4 +197,103 @@
  */
 - (void)log:(nonnull Class)class level:(DDLogLevel)level format:(nonnull NSString *)format, ...;
 
+/**
```
Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: All of these are class-level conveniences to make working at the class level cleaner.

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.m [hunk 3]
```diff
@@ -69,6 +69,10 @@ + (instancetype)sharedInstanceWithComponent:(NSString *)componentName {
     }
 }
 
++ (instancetype)sharedInstance {
+    return [self sharedInstanceWithComponent:kDefaultComponentName];
```
Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.m]: NB: `sharedDefaultInstance` could just go away with this implementation—it'd simply be the default implementation of `sharedInstance`.

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Util/SFSDKAnalyticsLogger.h [hunk 4]
```diff
@@ -29,65 +29,8 @@
 
 #import "SFSDKLogger.h"
 
-@interface SFSDKAnalyticsLogger : NSObject
+extern NSString * _Nonnull const kSFSDKAnalyticsComponentName;
 
-/**
```
Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Util/SFSDKAnalyticsLogger.h]: Basically, all of the component-specific code can simply be pulled up into the parent class, leaving only the component configuration.

File: libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h (hunk 2 above)
Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: I'd rather not have class-level conveniences, at least not for setting and getting log level. The behavior can be misleading to developers especially based on our 5.2 documentation. For example, a developer might use `[SFSDKSmartStoreLogger setLogLevel:...]` thinking that he/she is setting the log level for `SmartStore`, but it's actually being set for the default component. Same goes for any of the class level logging methods.

Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: They will be set for the proper component.  The key is the `sharedInstance` override.  Each of our established components (like `SFSDKAnalyticsLogger` is here) override this method, to return their own component.  Same for the log messages.

The only time that they'd go to the default is if you e.g. did `[SFSDKLogger setLogLevel:]`, which IMO is also intuitive—the default is simply the base class.  Thoughts?

Comment (bhariharan) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: You're right. My bad. That's why I shouldn't be doing code reviews at 6 A.M. with no sleep the previous night.

Looks good to me. 👍

Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: Cool.  I'll do the follow-on components in the next PR.

Comment (wmathurin) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: Interesting. That would not work in Java (you can't override static methods).
For a good read on why java is that way see [this](https://stackoverflow.com/questions/2223386/why-doesnt-java-allow-overriding-of-static-methods).

Comment (khawkins) [inline libs/SalesforceAnalytics/SalesforceAnalytics/Classes/Logger/SFSDKLogger.h]: I contemplated this when I was initially looking at solutions.  In a more classic OO design pattern, it seems that you'd separate your `LoggingConfiguration` class (which would be more of an instance class) from your `Logger` class (which would be more of a service class that owned `LoggingConfiguration` instances).  In the current design, those roles are both aggregated in the same class.

While I think I'd like the clean separation in that approach, it seemed like a bigger departure from the current implementation.  It's an idea I could poke at, though, if you guys think that would be better.

**Pull request merged**